pkg/microservice/policy/core/service/permission.go

```diff
@@ -74,3 +59,103 @@ func GetPermission(ns, uid string, log *zap.SugaredLogger) (map[string][]string,
+	// 1. get all policyBindings
+	policyBindings, err := ListPolicyBindings(projectName, uid, logger)
+	if err != nil {
+		logger.Errorf("ListPolicyBindings err:%s", err)
+		return nil, err
+	}
+	var policies []*Policy
+	for _, v := range policyBindings {
+		policy, err := GetPolicy(projectName, v.Policy, logger)
+		if err != nil {
+			logger.Warnf("GetPolicy err:%s", err)
+			continue
+		}
+		policies = append(policies, policy)
+	}
+
+	queryResourceSet := sets.NewString(resources...)
+	resourceM := make(map[string]sets.String)
+	for _, v := range resources {
+		resourceM[v] = sets.NewString()
+	}
+	for _, policy := range policies {
+		for _, rule := range policy.Rules {
+			if rule.Resources[0] == resourceType {
+				for _, resource := range rule.RelatedResources {
+					if queryResourceSet.Has(resource) {
+						resourceM[resource] = resourceM[resource].Insert(rule.Verbs...)
+					}
+				}
+			}
+
+		}
+	}
+	// 2. get all roleBindings
+	roleBindings, err := ListUserAllRoleBindings(projectName, uid)
+	if err != nil {
+		logger.Errorf("ListUserAllRoleBindings err:%s", err)
+		return nil, err
+	}
+	roles, err := ListUserAllRolesByRoleBindings(roleBindings)
+	if err != nil {
+		logger.Errorf("ListUserAllRolesByRoleBindings err:%s", err)
+		return nil, err
+	}
+	for _, role := range roles {
+		if role.Name == string(setting.SystemAdmin) || role.Name == string(setting.ProjectAdmin) {
+			for k, _ := range resourceM {
+				resourceM[k] = sets.NewString("*")
+			}
+			break
+		}
+		for _, rule := range role.Rules {
+			if rule.Resources[0] == resourceType && resourceType != string(config.ResourceTypeProduct) {
+				for k, v := range resourceM {
+					resourceM[k] = v.Insert(rule.Verbs...)
+				}
+			}
+			if (rule.Resources[0] == "Environment" || rule.Resources[0] == "ProductionEnvironment") && resourceType == string(config.ResourceTypeProduct) {
+				var rs []label.Resource
+				for _, v := range resources {
+					r := label.Resource{
+						Name:        v,
+						ProjectName: projectName,
+						Type:        string(config.ResourceTypeProduct),
+					}
+					rs = append(rs, r)
+				}
+
+				labelRes, err := label.New().ListLabelsByResources(label.ListLabelsByResourcesReq{rs})
+				if err != nil {
+					continue
+				}
+				for _, resource := range resources {
+					resourceKey := fmt.Sprintf("%s-%s-%s", config.ResourceTypeProduct, projectName, resource)
+					if labels, ok := labelRes.Labels[resourceKey]; ok {
+						for _, label := range labels {
+							if label.Key == "production" {
+								if rule.Resources[0] == "Environment" && label.Value == "false" {
+									resourceM[resource] = resourceM[resource].Insert(rule.Verbs...)
+								}
+								if rule.Resources[0] == "ProductionEnvironment" && label.Value == "true" {
+									resourceM[resource] = resourceM[resource].Insert(rule.Verbs...)
+								}
+							}
+						}
+					}
+				}
+
+			}
+		}
+	}
```

Comment:
is it possible to optimize the logic from line 109, there are too many ```for loop``` and ```if``` statements which makes the code very hard to read.

Reply:
not used